jsinterop-ts-defs-doclet/src/main/java/com/vertispan/tsdefs/TsDoclet.java

```diff
@@ -0,0 +1,254 @@
+              .filter(
+                  element ->
+                      element.getEnclosedElements().stream()
+                          .anyMatch(
+                              e ->
+                                  nonNull(e.getAnnotation(JsConstructor.class))
+                                      && !e.getModifiers().contains(Modifier.NATIVE)))
+              .collect(Collectors.toSet());
+
+      Set<Element> methodsParentElements =
+          environment.getIncludedElements().stream()
+              .filter(element -> isNull(element.getAnnotation(JsType.class)))
+              .filter(
+                  element ->
+                      element.getEnclosedElements().stream()
+                          .anyMatch(
+                              e ->
+                                  nonNull(e.getAnnotation(JsMethod.class))
+                                      && !e.getModifiers().contains(Modifier.NATIVE)))
+              .collect(Collectors.toSet());
+
+      Set<Element> propertiesParentElements =
+          environment.getIncludedElements().stream()
+              .filter(element -> isNull(element.getAnnotation(JsType.class)))
+              .filter(
+                  element ->
+                      element.getEnclosedElements().stream()
+                          .anyMatch(
+                              e ->
+                                  nonNull(e.getAnnotation(JsProperty.class))
+                                      && !e.getModifiers().contains(Modifier.NATIVE)))
+              .collect(Collectors.toSet());
+
+      Set<Element> eligibleElements = new HashSet<>();
+      eligibleElements.addAll(jsTypes);
+      eligibleElements.addAll(jsFunctions);
+      eligibleElements.addAll(constructorsParentElements);
+      eligibleElements.addAll(methodsParentElements);
+      eligibleElements.addAll(propertiesParentElements);
+
+      eligibleElements.stream()
+          .filter(
+              element ->
+                  isNull(element.getAnnotation(JsType.class))
+                      || !element.getAnnotation(JsType.class).isNative())
+          .forEach(
+              e -> {
+                try {
+                  new TypeVisitor(e, this).visit(tsModuleBuilder);
+                } catch (Exception exception) {
+                  messager()
+                      .printMessage(
+                          Diagnostic.Kind.ERROR, Formatting.formatException(exception), e);
+                  throw new RuntimeException(exception);
+                }
+              });
+      String moduleName =
+          environment.getIncludedElements().stream()
+              .filter(element -> nonNull(element.getAnnotation(TsModule.class)))
+              .findFirst()
+              .map(element -> element.getAnnotation(TsModule.class).value())
+              .orElse("types");
+
+      File outFile = Paths.get(outputDir, moduleName + ".d.ts").toFile();
+
+      outFile.getParentFile().mkdirs();
```

Comment:
IJ probably told you to test if this returned true... also consider the Files.createDirectories(outFile.parent()) or whatever and we'll get an IO exception instead (can group it in a bigger try/catch with the block below).